rclrs/src/node/service.rs

```diff
@@ -0,0 +1,185 @@
+use std::borrow::Borrow;
+use std::boxed::Box;
+use std::ffi::CString;
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use rosidl_runtime_rs::Message;
+
+use crate::node::publisher::MessageCow;
+
+use parking_lot::{Mutex, MutexGuard};
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_service_t {}
+
+/// Internal struct used by services.
+pub struct ServiceHandle {
+    handle: Mutex<rcl_service_t>,
+    node_handle: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ServiceHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_service_t> {
+        self.handle.lock()
+    }
+}
+
+impl Drop for ServiceHandle {
+    fn drop(&mut self) {
+        let handle = self.handle.get_mut();
+        let node_handle = &mut *self.node_handle.lock();
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_service_fini(handle as *mut _, node_handle as *mut _);
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Service structs
+/// See [`Service<T>`] for an example
+pub trait ServiceBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &ServiceHandle;
+    /// Tries to take a new request and run the callback with it.
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+type ServiceCallback<Request, Response> =
+    Box<dyn Fn(&rmw_request_id_t, Request) -> Response + 'static + Send>;
+
+/// Main class responsible for responding to requests sent by ROS clients
```

Comment:
```suggestion
/// Main class responsible for responding to requests sent by ROS clients.
```

Reply:
Optionally, you could add some more detail on services. See e.g. the docs for `Subscription<T>`. The same for clients.